DLCS.Repository/Spaces/SpaceRepository.cs

```diff
@@ -30,31 +37,144 @@ public SpaceRepository(DlcsContext dlcsContext)
+            return space;
         }
 
-        public Task<Space?> GetSpace(int customerId, string name, CancellationToken cancellationToken)
+        public async Task<Space> CreateSpace(int customer, string name, string? imageBucket, 
+            string[]? tags, string[]? roles, int? maxUnauthorised, CancellationToken cancellationToken)
         {
-            throw new System.NotImplementedException();
+            int newModelId = await GetIdForNewSpace(customer);
+            
+            var space = new Space
+            {
+                Customer = customer,
+                Id = newModelId,
+                Name = name,
+                Created = DateTime.UtcNow,
+                ImageBucket = imageBucket,
+                Tags = tags ?? Array.Empty<string>(),
+                Roles = roles ?? Array.Empty<string>(),
+                MaxUnauthorised = maxUnauthorised ?? -1
+            };
+
+            await dlcsContext.Spaces.AddAsync(space, cancellationToken);
+            await entityCounterRepository.Create(customer,  "space-images", space.Id.ToString());
+            await dlcsContext.SaveChangesAsync(cancellationToken);
+            return space;
         }
+        
+        private async Task<int> GetIdForNewSpace(int requestCustomer)
+        {
+            int newModelId;
+            Space existingSpaceInCustomer;
+            do
+            {
+                var next = await entityCounterRepository
+                    .GetNext(requestCustomer, "space", requestCustomer.ToString());
+                newModelId = Convert.ToInt32(next);
+                existingSpaceInCustomer = await dlcsContext.Spaces
+                    .SingleOrDefaultAsync(s => s.Id == newModelId && s.Customer == requestCustomer);
+            } while (existingSpaceInCustomer != null);
 
-        public Task<Space> CreateSpace(int customer, string name, string? imageBucket, string[]? tags, string[]? roles, int? maxUnauthorised,
-            CancellationToken cancellationToken)
+            return newModelId;
+        }
+        
+
+        private async Task<Space?> GetSpaceInternal(int customerId, int spaceId, 
+            CancellationToken cancellationToken, string? name = null)
         {
-            throw new System.NotImplementedException();
+            Space space;
+            if (name != null)
+            {
+                space = await dlcsContext.Spaces
+                    .Where(s => s.Customer == customerId)
+                    .SingleOrDefaultAsync(s => s.Name == name, cancellationToken: cancellationToken);
+            }
+            else
+            {
+                space = await dlcsContext.Spaces.AsNoTracking().SingleOrDefaultAsync(s =>
+                    s.Customer == customerId && s.Id == spaceId, cancellationToken: cancellationToken);
+            }
+            var counter = await dlcsContext.EntityCounters.AsNoTracking().SingleOrDefaultAsync(ec =>

```

Comment:
Given it's an additional DB hit and not always required, is it worth having `bool withImageCount` param on this method?

Then, before getting `counter` you could have `if (space == null || !withImageCount) return space` to avoid extra DB hit.